### DISCUSSION TITLE
Text Source

### DISCUSSION THREAD

**Discussion by lupin3rd** [brunoherbelin/vimix#111] [Ideas]
A discussion about text source features.
Resolume has a very good text instrument that present those features:
![01](https://github.com/brunoherbelin/vimix/assets/1455316/5a0950bf-00ff-4223-8a9d-f8e5ee49f610)

Other interesting features:
- load a text file
- set text field and other options by OSC
- customizable mode to show text (one char at time, one word at time); in this way we can sync music with an OSC message that go to next char or to next word in the text field or in the text file

**Comment by brunoherbelin** [comment]
Thanks for opening the discussion ! Good that I finally address this and add a long awaited text source !

The text will be rendered with gstreamer [textrender](https://gstreamer.freedesktop.org/documentation/pango/textrender.html?gi-language=c) : all properties for font description and layout will be exposed in properties of the source.

The text source will take  a string as input:

- if the string is a filename, if the file is found and readable, then the content of the file will be used:
   - if the file is a subtitle format, then the [subparse](https://gstreamer.freedesktop.org/documentation/subparse/subparse.html?gi-language=c#subparse) plugin will be used and the text will change in time accordingly
    - if the file is a raw text file, the entire content will be shown as one string
- if the string is not a filename, then the content of the string will be shown

I will also expose the change of the string on the OSC API, e.g. 
```
/vimix/current/content s 'hello vimix'
```
This way, you can send from OSC a new string at anytime to do the animation you want :smile: 

More advanced or custom text features would be beyond what vimix could do. However, then thinking the problem differently and preparing all rendered frames with the animated text and creating an image sequence source (that can be played, stopped, seeked, play next frame, etc.) would be an alternative with complete freedom!

**Comment by brunoherbelin** [comment]
Okay, let me know if you can test the new Text source in Beta 58371c36d3175d16e5fdc9680a2c728a0b9a9f35

Here is how it should look like;

![image](https://github.com/brunoherbelin/vimix/assets/60393189/81291c13-d550-4748-a24b-698acd590b93)

It uses the gst [textoverlay](https://gstreamer.freedesktop.org/documentation/pango/textoverlay.html?gi-language=c) plugin with Pango [markup](https://docs.gtk.org/Pango/pango_markup.html) for rich test and for [font description](https://docs.gtk.org/Pango/type_func.FontDescription.from_string.html).

**Comment by brunoherbelin** [comment]
Example animation with subtitle : save this [hello.txt](https://github.com/brunoherbelin/vimix/files/13334975/hello.txt) file as `hello.srt` and open from the Text source creation;

![image](https://github.com/brunoherbelin/vimix/assets/60393189/fd52a097-ee04-4d2e-b75e-7b3ca11007b6)

**Comment by lupin3rd** [comment]
Hi Bruno!
It's very cool :-D and very complete!
I tested all functions on GUI and works very fine!
Tomorrow i will do some OSC tests !

**Reply by brunoherbelin** [reply]
Good :smiley: 
Please note the OSC target is `contents`
e.g.,
```
oscsend localhost 7000 /vimix/#1/contents s "we <b>&lt;3</b> <i>vimix</i>"
```

**Comment by kubriel** [comment]
yep, seems to be working nicely, thanks!

**Comment by lupin3rd** [comment]
With OSC message i can't send an empty message, i try some solutions but no way. It's usefull if i want to clean text.

**Comment by brunoherbelin** [comment]
Accepting empty string in Beta e42afcb4343735d7eb90e68856792dcf3ebf0a83
